File: 20 Valid Parentheses/20 LL.cpp
#include <vector>
#include <iostream>
using namespace std;

/*
https://leetcode.com/problems/valid-parentheses/

Given a string containing just the characters '(', ')', '{', '}', '[' and ']',
determine if the input string is valid.

The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
*/

class Solution
{
public:
	bool isValid(string s)
	{
		vector<char> stack(1, '*');
		string::iterator data = s.begin();
		for (; data != s.end(); ++data)
		{
			switch (data[0])
			{
			case '(':
			case '{':
			case '[':
				stack.push_back(data[0]);
				break;
			case ')':
				if (stack.back() != '(') return false;
				stack.pop_back();
				break;
			case '}':
				if (stack.back() != '{') return false;
				stack.pop_back();
				break;
			case ']':
				if (stack.back() != '[') return false;
				stack.pop_back();
				break;
			default:
				return false;
			}
		}
		return (stack.size() == 1);
	}
};

/*
idea: using stack to remember pairs
complexity: Time O(N)
*/

int main(void)
{
	Solution engine;
	cout << engine.isValid("([{}])") << '\n';
	return 0;
}

File: 20 Valid Parentheses/20 LL.py
class Solution(object):
	def isValid(self, s):
		stack = ['*']
		for c in s:
			if c == '{' or c == '[' or c == '(':
				stack.append(c)
			elif c == '}':
				if stack[-1] != '{':
					return False
				stack.pop()
			elif c == ']':
				if stack[-1] != '[':
					return False
				stack.pop()
			elif c == ')':
				if stack[-1] != '(':
					return False
				stack.pop()
			else:
				return False
		return len(stack) == 1

engine = Solution()
print engine.isValid("([)]")